drivers/node/vsphere/vsphere.go

```diff
@@ -0,0 +1,248 @@
+	v.SSH.Init()
+
+	username := os.Getenv(vsphereUname)
+	if len(username) == 0 {
+		v.vsphereUsername = DefaultUsername
+	} else {
+		v.vsphereUsername = username
+	}
+
+	v.vspherePassword = os.Getenv(vspherePwd)
+	if len(v.vspherePassword) == 0 {
+		return fmt.Errorf("Vsphere password not provided as env var: %s", vspherePwd)
+	}
+
+	v.vsphereHostIP = os.Getenv(vsphereIP)
+	if len(v.vsphereHostIP) == 0 {
+		return fmt.Errorf("Vsphere host IP not provided as env var: %s", vsphereIP)
+	}
+	err := v.connect()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TestConnection tests the connection to the given node
+func (v *vsphere) TestConnection(n node.Node, options node.ConnectionOpts) error {
+	var err error
+	logrus.Infof("Testing vsphere driver connection by checking state of the VMs in the vsphere")
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Failed to get VM: %s", n.Name)
+	}
+	vm := vmMap[n.Name]
+
+	t := func() (interface{}, bool, error) {
+		powerState, err := vm.PowerState(v.ctx)
+		if err != nil || powerState != types.VirtualMachinePowerStatePoweredOn {
+			return nil, true, &node.ErrFailedToRebootNode{
+				Node:  n,
+				Cause: fmt.Sprintf("Failed to test connection to VM: %s Current Status: %v, error: %v", vm.Name(), powerState, err),
+			}
+		}
+		return nil, false, nil
+	}
+	if _, err := task.DoRetryWithTimeout(t, VMReadyTimeout, VMRetryInterval); err != nil {
+		return err
+	}
+	return err
+}
+
+func (v *vsphere) connect() error {
+
+	login := fmt.Sprintf("%s%s:%s@%s/sdk", Protocol, v.vsphereUsername, v.vspherePassword, v.vsphereHostIP)
+	logrus.Infof("Logging in to ESXi using: %s", login)
+
+	u, err := url.Parse(login)
+	if err != nil {
+		return fmt.Errorf("error parsing url %s", login)
+	}
+
+	v.ctx, v.cancel = context.WithCancel(context.Background())
+	//defer cancel()
+
+	c, err := govmomi.NewClient(v.ctx, u, true)
+	if err != nil {
+		return fmt.Errorf("logging in error: %s", err.Error())
+	}
+	logrus.Infof("Log in successful to vsphere:  %s:\n", v.vsphereHostIP)
+
+	f := find.NewFinder(c.Client, true)
+
+	// Find one and only datacenter
+	dc, err := f.DefaultDatacenter(v.ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to find data center: %v", err)
+	}
+
+	// Make future calls local to this datacenter
+	f.SetDatacenter(dc)
+
+	// Find virtual machines in datacenter
+	vms, err := f.VirtualMachineList(v.ctx, "*")
+	if err != nil {
+		return fmt.Errorf("Failed to find any virtual machines on %s: %v", v.vsphereHostIP, err)
+	}
+
+	nodes := node.GetNodes()
+	for _, vm := range vms {
+		for _, n := range nodes {
+			if vm.Name() == n.Name {
+				if _, ok := vmMap[vm.Name()]; !ok {
+					vmMap[vm.Name()] = vm
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// RebootVM reboots vsphere VM
+func (v *vsphere) RebootNode(n node.Node, options node.RebootNodeOpts) error {
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Could not fetch VM for node: %s", n.Name)
+	}
+
+	vm := vmMap[n.Name]
```

Comment:
I would implement getVmByName and get a VM from vSphere in runtime, then you don't need to keep vmMap

Reply:
I would say, it makes sense to have an in-memory object that contains only the VMs being used in this test rather than querying the vSphere every time. Currently there is a similar data structure nodeMap being used for the same reason.